modules/compute/vm-instance/main.tf

```diff
@@ -128,9 +115,13 @@ resource "google_compute_instance" "compute_vm" {
   labels = local.labels
 
   boot_disk {
-    source      = google_compute_disk.boot_disk[count.index].self_link
-    device_name = google_compute_disk.boot_disk[count.index].name
```

Comment:
Shall we keep it `"${local.resource_prefix}-boot-disk-${count.index}"` ?

Reply:
- newly created VMs will have disks with the same name as the VM (similar, but not identical)
- if one deploys the blueprint above using develop and then re-deploys off this PR, it will fail to destroy the attached disk:

  ```text
  Display full proposed changes, Apply proposed changes, Stop and exit, Continue without applying? [d,a,s,c]: a
  2023/05/23 10:43:33 running terraform apply on group simple/primary
  module.simplevm.google_compute_disk.boot_disk[0]: Destroying... [id=projects/toolkit-demo-zero-e913/zones/us-central1-b/disks/simple-boot-disk-0]

  Error: Error detaching disk simple-boot-disk-0 from instance toolkit-demo-zero-e913/us-central1-b/simple-0: googleapi: Error 400: Invalid resource usage: 'To detach the boot disk, the instance must be in TERMINATED state.'., invalidResourceUsage
  ```

So, it will fail but in a way that does not destroy existing infrastructure. [moved] blocks cannot address this because we are intentionally no longer managing a separate `google_compute_disk` resource. My opinion is that separate management of the disk is sufficiently counterproductive that removing it is worth it and that we should advise users to run:

```shell
❯❯❯ terraform state rm module.simplevm.google_compute_disk.boot_disk
Removed module.simplevm.google_compute_disk.boot_disk[0]
Successfully removed 1 resource instance(s).
❯❯❯ terraform plan
module.simplevm.module.netstorage_startup_script.random_id.resource_name_suffix: Refreshing state... [id=1SOGGg]
module.simplevm.data.google_compute_image.compute_image: Reading...
module.network1.data.google_compute_subnetwork.primary_subnetwork: Reading...
module.network1.data.google_compute_network.vpc: Reading...
module.simplevm.module.netstorage_startup_script.google_storage_bucket.configs_bucket[0]: Refreshing state... [id=simple-startup-scripts-d523861a]
module.simplevm.module.netstorage_startup_script.google_storage_bucket_object.scripts["passed_startup_script.sh"]: Refreshing state... [id=simple-startup-scripts-d523861a-passed_startup_script.sh-5617]
module.network1.data.google_compute_subnetwork.primary_subnetwork: Read complete after 0s [id=projects/toolkit-demo-zero-e913/regions/us-central1/subnetworks/default]
module.network1.data.google_compute_network.vpc: Read complete after 0s [id=projects/toolkit-demo-zero-e913/global/networks/default]
module.simplevm.data.google_compute_image.compute_image: Read complete after 1s [id=projects/cloud-hpc-image-public/global/images/hpc-centos-7-v20230419]
module.simplevm.google_compute_instance.compute_vm[0]: Refreshing state... [id=projects/toolkit-demo-zero-e913/zones/us-central1-b/instances/simple-0]

No changes. Your infrastructure matches the configuration.
```

In this case, the `google_compute_instance` is "unharmed" because it detects the boot disk attached to it already (even if it has the "wrong" name) and does not see a need to re-create the resource.

I'll investigate a bit to see if there is anything more automatic that might help us here (e.g. precondition/postcondition).

[moved]: https://developer.hashicorp.com/terraform/language/modules/develop/refactoring

Reply:
@tpdownes, under what conditions does the user get the error you describe: `Error detaching disk simple-boot-disk-0 from instance`? 

Does this happen even if I don't change anything on the vm-instance settings?  
Does this happen whether or not `auto_delete_boot_disk` is set to `true` or `false`?

I suspect the answer is yes to both.

Reply:
`yes` to both on HEAD of PR.